Napari gui part4 (#30)
BiAPoL/blog: marcelo_zoccoler/entry_user_interf4/Readme.md

* [Creating standalone GUIs (Part 2)](https://biapol.github.io/blog/johannes_mueller/entry_user_interf2#creating-advanced-standalone-guis)
* [Creating GUIs for napari (Part 3)](https://biapol.github.io/blog/marcelo_zoccoler/entry_user_interf3#creating-advanced-guis-for-napari)
* [Turning napari GUI into plugins (Part 4)](https://biapol.github.io/blog/marcelo_zoccoler/entry_user_interf4#turning-napari-gui-into-plugins)

# Turning napari GUI into plugins
The [previous entry](https://biapol.github.io/blog/marcelo_zoccoler/entry_user_interf3#creating-advanced-guis-for-napari) showed you three different ways to create GUIs and embbed them into napari locally. This forth and last part will teach you how to transform them into napari plugins! 🏝️ 🚀

## Table of contents
* [Creating a plugin template with cookiecutter](#creating-a-plugin-template-with-cookiecutter)
* [Adding your local repository to Github](#adding-your-local-repository-to-github)
* [Putting your GUI into the template](#putting-your-gui-into-the-template)
* [Updating Github repository](#updating-github-repository)
* [Publishing your plugin](#publishing-your-plugin)

> ## Creating a plugin template with cookiecutter

Reviewer: ```suggestion
## Creating a plugin from a template using the cookiecutter
```

Well, first things first: [create a new conda environment](https://biapol.github.io/blog/johannes_mueller/entry_user_interf2/Readme.md#creating-your-environment). ("Another one?" 😩) Yes, better safe than sorry 😉. Don't forget to activate it.

> Napari website has a complete [tutorial](https://napari.org/plugins/stable/for_plugin_developers.html) for creating plugins, and I would say the easiest way is by using [cookiecutter](https://napari.org/plugins/stable/for_plugin_developers.html#cookiecutter-template).

Reviewer: ```suggestion
The napari website has a complete [tutorial](https://napari.org/plugins/stable/for_plugin_developers.html) for creating plugins, and I would say the easiest way is by using [cookiecutter](https://napari.org/plugins/stable/for_plugin_developers.html#cookiecutter-template).
```

> So, let's install and run cookiecutter:

Reviewer: ```suggestion
So, let's install and run the cookiecutter as suggested in the [template readme](https://github.com/napari/cookiecutter-napari-plugin):
```

        super().__init__(
          flood_fgui,
          call_button=True,
          layout='vertical',
          param_options={'delta':
                             {'label': 'Temperature Increase (Δ°C):', 
                              'min': 0, 'max' : 3, 'step': 0.1},
                        'new_level':
                            {'label':'Sea Level (dm):', 'widget_type':'Slider',
                             'min': 0, 'max' : 255}}
        )
        
    def __call__(self):
        label_image = super().__call__()
        new_level = round(label_image[1]['metadata']['new_level'])
        self.new_level.value = new_level

@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    # you can return either a single widget, or a sequence of widgets
    return [ExampleQWidget, example_magic_widget]
```

Last thing is to replace `napari_experimental_provide_dock_widget` outputs by our own. In our case, it should look like this:

```Python
@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    # you can return either a single widget, or a sequence of widgets
    return [Qt_Designer_flood, flood_magic_factory, FunctionGui_flood]
```

> If you now call `napari` again, your plugin, along with each of the 3 sub-menu versions, should be available and working!

Reviewer: ```suggestion
If you now start `napari` from the terminal again, your plugin, along with each of the 3 sub-menu versions, should be available and working!
```

![](images/flood_plugin_v1.png)

> Obs: remember we wrote our qt_designer version to work with a specific layer name (`napari_island`). There are other ways of selecting the appropriated layer, like by order, type, etc. You can have access to available layers through the viewer instance `viewer.layers`.

Reviewer: What does "Obs" mean? Can you write it out? (-:

Author: ooops... I thought "Obs" was standard for an additional "observation". I think PS was the closest acronym I was looking for.
I will replace for 'Remider', I think it's best.

## Updating Github repository

We have our updates locally in our separated branch. Let's push them into our Github repository. If you open Github Desktop again now, you should see all the changes we made listed in the left of the screen. Write a description to these changes (step 1.), click on 'commit to 'branch name'' (step 2.), and then publish your new branch (step 3.):

![](images/flood_plugin_commit.png)

After that, click on the 'Create Pull Request' button, and you will be prompted to a browser tab with your github online repository. There, click on 'Create Pull Request'. 

![](images/flood_plugin_PR.png)

After the page updates itself, click on 'Merge Pull Request' and confirm it if required.

![](images/flood_plugin_merge.png)

At the end, you should see the image below, which means our online repository is now updated with your new code!

![merge_done](https://user-images.githubusercontent.com/26173597/145111544-dea73c05-a8a5-47bb-85c4-d9431ca0de0e.png)

> ## Publishing your plugin

Reviewer: ```suggestion
## Publishing your plugin
After you developed your plugin to a point where you think it's ready to share it with the world, it's time so submit it to the python package index (PyPI).
```

Publishing the plugin on [PyPi](https://pypi.org/) may sound like a big ordeal, but is, in fact, very simple. To do so, first create an account on [PyPi](https://pypi.org/account/register/). After this is done, open an anaconda command prompt and `cd` into your project repository. 

Next, you need to create the necessary packaging information from your sourcefiles. You can do so by either creating a *source distribution* of your package with 
```bash
python setup.py sdist
```

or a *wheel* using
```bash
python setup.py bdist_wheel
```

> The difference between both is, in short, that the source distribution provides, as the name suggests, the soure files that can be downloaded and have to be compiled upon installation of the package. Python wheels, on the contrary, come pre-built, which leads to faster installation. This can be convenient for large packages. For a more in-depth on the advantages of either strategy, see [this blog](https://medium.com/ochrona/understanding-python-package-distribution-types-25d53308a9a). For simple projects, there is no damage in simply providing both. 

Reviewer: ```suggestion
The difference between both is, in short, that the source distribution provides, as the name suggests, the source files that can be downloaded and have to be compiled upon installation of the package. Python wheels, on the contrary, come pre-built, which leads to faster installation. This can be convenient for large packages. For a more in-depth on the advantages of either strategy, see [this blog](https://medium.com/ochrona/understanding-python-package-distribution-types-25d53308a9a). For simple projects, there is no damage in simply providing both. 
```
